File: Rezoom.Test/TestExceptionCatching.fs
module Rezoom.Test.TestExceptionCaching
open System
open Rezoom
open NUnit.Framework
open FsUnit

[<Test>]
let ``simple catch`` () =
    {   Task = fun () ->
            plan {
                try
                    explode "fail"
                    return 2
                with
                | ArtificialFailure "fail" -> return 1
            }
        Batches = []
        Result = Good 1
    } |> test
        
[<Test>]
let ``bound catch`` () =
    {   Task = fun () ->
            plan {
                try
                    let! x = send "x"
                    explode "fail"
                    return 2
                with
                | ArtificialFailure "fail" -> return 1
            }
        Batches = [["x"]]
        Result = Good 1
    } |> test

[<Test>]
let ``simple failing prepare`` () =
    {   Task = fun () ->
            plan {
                try
                    let! x = failingPrepare "fail" "x"
                    return 2
                with
                | PrepareFailure "fail" -> return 1
            }
        Batches = []
        Result = Good 1
    } |> test

[<Test>]
let ``simple failing retrieve`` () =
    {   Task = fun () ->
            plan {
                try
                    let! x = failingRetrieve "fail" "x"
                    return 2
                with
                | RetrieveFailure "fail" -> return 1
            }
        Batches = [["x"]]
        Result = Good 1
    } |> test

[<Test>]
let ``concurrent catching`` () =
    let catching query =
        plan {
            let guid = Guid.NewGuid().ToString()
            try
                let! x = failingRetrieve guid query
                return x
            with
            | RetrieveFailure msg when msg = guid -> return "bad"

        }
    let good query =
        plan {
            let! result = send query
            return result
        }
    {   Task = fun () ->
            plan {
                let! x, y, z =
                    catching "x", catching "y", good "z"
                return x + y + z
            }
        Batches =
            [   [ "x"; "y"; "z" ]
            ]
        Result = Good "badbadz"
    } |> test

[<Test>]
let ``concurrent loop catching`` () =
    let catching query =
        plan {
            let guid = Guid.NewGuid().ToString()
            try
                let! x = failingRetrieve guid query
                return ()
            with
            | RetrieveFailure msg when msg = guid -> return ()

        }
    let good query =
        plan {
            let! result = send query
            return ()
        }
    {   Task = fun () ->
            plan {
                for q in batch ["x"; "y"; "z"] do
                    if q = "y" then
                        do! good q
                    else
                        do! catching q
                return ()
            }
        Batches =
            [   [ "x"; "y"; "z" ]
            ]
        Result = Good ()
    } |> test

[<Test>]
let ``concurrent non-catching`` () =
    let notCatching query =
        plan {
            let! x = failingRetrieve "fail" query
            return x
        }
    let good query =
        plan {
            let! result = send query
            let! next = send "jim"
            return result + next
        }
    {   Task = fun () ->
            plan {
                let! x, y, z =
                    notCatching "x", notCatching "y", good "z"
                return x + y + z
            }
        Batches =
            [   [ "x"; "y"; "z" ]
            ]
        Result = Bad (fun ex -> true)
    } |> test

File: src/Rezoom.CS.Test/TestExceptionCatching.cs
using System;
using NUnit.Framework;
using Rezoom;
using Rezoom.CS;
using static Rezoom.CS.Test.Helpers;

namespace Rezoom.CS.Test;

[TestFixture]
public class TestExceptionCatching
{
    [Test]
    public void Simple_catch()
    {
        async Plan<int> Task()
        {
            try
            {
                Explode("fail");
                return 2;
            }
            catch (ArtificialFailure e) when (e.Message == "fail")
            {
                return 1;
            }
        }
        TestRunner.Run(Task, System.Array.Empty<string[]>(), ExpectedResult<int>.Ok(1));
    }

    [Test]
    public void Bound_catch()
    {
        async Plan<int> Task()
        {
            try
            {
                var x = await Send("x");
                Explode("fail");
                return 2;
            }
            catch (ArtificialFailure e) when (e.Message == "fail")
            {
                return 1;
            }
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x" } },
            ExpectedResult<int>.Ok(1));
    }

    [Test]
    public void Simple_failing_prepare()
    {
        async Plan<int> Task()
        {
            try
            {
                var x = await FailingPrepare("fail", "x");
                return 2;
            }
            catch (PrepareFailure e) when (e.Message == "fail")
            {
                return 1;
            }
        }
        TestRunner.Run(Task, System.Array.Empty<string[]>(), ExpectedResult<int>.Ok(1));
    }

    [Test]
    public void Simple_failing_retrieve()
    {
        async Plan<int> Task()
        {
            try
            {
                var x = await FailingRetrieve("fail", "x");
                return 2;
            }
            catch (RetrieveFailure e) when (e.Message == "fail")
            {
                return 1;
            }
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x" } },
            ExpectedResult<int>.Ok(1));
    }

    [Test]
    public void Concurrent_catching()
    {
        async Plan<string> Catching(string query)
        {
            var guid = Guid.NewGuid().ToString();
            try
            {
                return await FailingRetrieve(guid, query);
            }
            catch (RetrieveFailure e) when (e.Message == guid)
            {
                return "bad";
            }
        }
        async Plan<string> Good(string query) => await Send(query);

        async Plan<string> Task()
        {
            var (x, y, z) = await Plans.Tuple(Catching("x"), Catching("y"), Good("z"));
            return x + y + z;
        }

        TestRunner.Run(
            Task,
            new[] { new[] { "x", "y", "z" } },
            ExpectedResult<string>.Ok("badbadz"));
    }

    [Test]
    public void Concurrent_loop_catching()
    {
        async Plan<int> Catching(string query)
        {
            var guid = Guid.NewGuid().ToString();
            try
            {
                _ = await FailingRetrieve(guid, query);
                return 0;
            }
            catch (RetrieveFailure e) when (e.Message == guid)
            {
                return 0;
            }
        }
        async Plan<int> Good(string query) { _ = await Send(query); return 0; }

        async Plan<int> Task()
        {
            await Plans.ForBatch(new[] { "x", "y", "z" }, async q =>
            {
                if (q == "y") return await Good(q);
                return await Catching(q);
            });
            return 0;
        }

        TestRunner.Run(
            Task,
            new[] { new[] { "x", "y", "z" } },
            ExpectedResult<int>.Ok(0));
    }

    [Test]
    public void Concurrent_non_catching()
    {
        async Plan<string> NotCatching(string query) => await FailingRetrieve("fail", query);
        async Plan<string> Good(string query)
        {
            var result = await Send(query);
            var next = await Send("jim");
            return result + next;
        }
        async Plan<string> Task()
        {
            var (x, y, z) = await Plans.Tuple(NotCatching("x"), NotCatching("y"), Good("z"));
            return x + y + z;
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x", "y", "z" } },
            ExpectedResult<string>.Fail());
    }
}
